### PR TITLE
Adding master password decrypting

### LINKED ISSUE #4589: Authentication failure when using password encryption through Maven `settings-security.xml`
Hello,

I have an issue running custom recipie for `org.openrewrite.maven.ChangeParentPom`

Here is my configuration:
A maven, multi-module project. I want to update the version of the parent of the root's pom.xml file.

Here is my configuration in rules.yml file:
```
type: specs.openrewrite.org/v1beta/recipe
name: my.test.UpgradeParent
displayName: Change Maven dependency example
recipeList:
  - org.openrewrite.maven.ChangeParentPom:
      oldGroupId: com.my.group
      oldArtifactId: my-artifact
      newVersion: 1.0.4
```

In my pom.xml I have:

```
<project>
    <modelVersion>4.0.0</modelVersion>

    <parent>
        <groupId>com.my.group</groupId>
        <artifactId>my-artifact</artifactId>
        <version>1.107</version>
    </parent>

    <build>
        <plugins>
            <plugin>
                <groupId>org.openrewrite.maven</groupId>
                <artifactId>rewrite-maven-plugin</artifactId>
                <version>5.42.2</version>
                <configuration>
                    <configLocation>rules.yml</configLocation>
                    <activeRecipes>
                        <recipe>my.test.UpgradeParent</recipe>
                    </activeRecipes>
                    <failOnInvalidActiveRecipes>true</failOnInvalidActiveRecipes>
                </configuration>
            </plugin>
        </plugins>
    </build>
...

```

But when I run it with rewrite:run or rewrite:dryRun, it doesn't match my parent pom GAV and don't update it...

```
[INFO] --- rewrite:5.42.2:dryRun (default-cli) @ XXXXX-module ---
[INFO] Using active recipe(s) [com.bnppf.UpgradeToSf5]
[INFO] Using active styles(s) []
[INFO] Validating active recipes...
[INFO] Project [XXXXX] Resolving Poms...
[INFO] Project [XXXXX] Parsing source files
[INFO] Project [XXXXX-module1] Parsing source files
[INFO] Project [XXXXX-module2] Parsing source files
[INFO] Project [XXXXX-module3] Parsing source files
[INFO] Project [XXXXX-module4] Parsing source files
[INFO] Project [XXXXX-module5] Parsing source files
[INFO] Running recipe(s)...
[INFO] Applying recipes would make no changes. No patch file generated.
```

I tried to specify newArtifactId, newGroupId, but it changes nothing.... What am I doing wrong?
Any help will be highly appreciated...

Thank you very much in advance!

I am using:

- Maven plugin v5.42.2
- Java version: 1.8 (and 11 and 17)
- Maven version: 3.9.9

### PR DESCRIPTION
**Pull request by svaningelgem**
## What's changed?
Adding the ability for openrewrite to read encrypted passwords. (maven-feature as described here: 
https://maven.apache.org/guides/mini/guide-encryption.html

## What's your motivation?
We can't use this library on our internal Nexus server as our passwords are encrypted.

## Anything in particular you'd like reviewers to focus on?
For now this is a draft as I don't know what to add/change. Or maybe you can pick it up and whip it into shape? The heavy lifting for the decryption is already done after all ;-).

## Have you considered any alternatives or workarounds?
No alternatives exist.

## Any additional context
- Closes #4589

### Checklist
- [ ] I've added unit tests to cover both positive and negative cases
- [ ] I've read and applied the [recipe conventions and best practices](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices)
- [ ] I've used the IntelliJ IDEA auto-formatter on affected files

### CODE REVIEW

File: rewrite-maven/src/main/java/org/openrewrite/maven/MavenSettings.java [hunk 1]
```diff
@@ -113,6 +127,99 @@ public MavenSettings(@Nullable String localRepository, @Nullable Profiles profil
                 .orElse(installSettings);
     }
 
+    private byte[] extractPassword(@NotNull String pwd) {
```
[the remark this answers is not in the document]
Comment (timtebeek) [reply rewrite-maven/src/main/java/org/openrewrite/maven/MavenSettings.java:130]: As a bit of context: we use a package level non-null annotation, so it's not needed to add this to parameters.

[the remark this answers is not in the document]
Comment (svaningelgem) [reply rewrite-maven/src/main/java/org/openrewrite/maven/MavenSettings.java]: Thanks, fixed this

**Comment by svaningelgem** [comment]
@timtebeek : could you list all the changes that are necessary?

**Comment by timtebeek** [comment]
Looking good already @svaningelgem ! Could you remove the `ChangeParentPom2Test` and replace it with a test that has both a `security-settings.xml` and `settings.xml`, based on [the examples here](https://maven.apache.org/guides/mini/guide-encryption.html), and verifies that the decrypted values match their actual values after calling `maybeDecryptPasswords`?

**Comment by svaningelgem** [comment]
Thanks for the assist. I'll handle the hint you gave me

**Comment by svaningelgem** [comment]
Hello @timtebeek , I added the tests now. Could you have a look if they're decent enough and/or where I would need to make changes?

Thanks

**Comment by svaningelgem** [comment]
👍 Thanks @timtebeek for the assist ! :-)

**Review by timtebeek — APPROVED** [review]
Thanks a lot for the hard work here @svaningelgem ! I've only added some small changes to move things around and support relocation; apart from that very glad to see you tackled the decryption & tests. Hope this helps you all there!

**Comment by svaningelgem** [comment]
Yes, I will try to test it out at work tomorrow. I hope this solves the issue.

**Comment by sruffatti** [comment]
@svaningelgem This change is not working for me. 

my master password is in a file called settings-security.xml in my .m2 directory 

It looks like this
```
<settingsSecurity>
	<master>{my-encrypted-password}</master>
</settingsSecurity>
```

The exception is below.
```
java.lang.NegativeArraySizeException: -75
	at org.openrewrite.maven.MavenSecuritySettings.decrypt(MavenSecuritySettings.java:175)
	at org.openrewrite.maven.MavenSettings.lambda$maybeDecryptPasswords$4(MavenSettings.java:151)
	at org.openrewrite.internal.ListUtils.map(ListUtils.java:243)
	at org.openrewrite.internal.ListUtils.map(ListUtils.java:265)
	at org.openrewrite.maven.MavenSettings.maybeDecryptPasswords(MavenSettings.java:150)
	at org.openrewrite.maven.MavenSettings.readMavenSettingsFromDisk(MavenSettings.java:129)
	at com.org.engg.openrewrite.config.OrgRewriteTest.defaultExecutionContext(OrgRewriteTest.java:22)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:214)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:132)
	at org.openrewrite.test.RewriteTest.rewriteRun(RewriteTest.java:127)
```

**Comment by svaningelgem** [comment]
Hi Sean,

What I would suggest is that you check out the main branch and debug with your settings file.
It is very possible that it won't work with every file as I skipped some stuff.
The current code is proven to work with my file and password, but likely yours is longer/shorter/more variation/... so it triggers a side condition.

Alternatively, you can send me your settings & settings security file (check my profile to see how to reach out). Watch out. This will expose your password to me (but I promise to delete it once I'm done with it -- and my memory is worse than a goldfish ;)).

Grtz,
Steven

**Pull request merged**